tachyon/math/finite_fields/binary_prime_field.h

```diff
@@ -0,0 +1,223 @@
+  constexpr explicit PrimeField(Type value) : value_(value) {
+    if constexpr (kBits <= 8) {
+      DCHECK_LE(value_, GetMax());
+    }
+  }
+  constexpr explicit PrimeField(BigInt<N> value) : PrimeField(value[0]) {
+    if constexpr (kBits <= 8) {
+      DCHECK_LE(value[0], GetMax());
+    }
+  }
+  constexpr PrimeField(const PrimeField& other) = default;
+  constexpr PrimeField& operator=(const PrimeField& other) = default;
+  constexpr PrimeField(PrimeField&& other) = default;
+  constexpr PrimeField& operator=(PrimeField&& other) = default;
+
+  constexpr static PrimeField Zero() { return PrimeField(); }
+  constexpr static PrimeField One() { return PrimeField(1); }
+
+  static PrimeField Random() {
+    return PrimeField(
+        base::Uniform(base::Range<Type, true, true>::Until(GetMax())));
+  }
+
+  static std::optional<PrimeField> FromDecString(std::string_view str) {
+    using MaybePromotedType = std::conditional_t<(kBits <= 32), uint32_t, Type>;
+    MaybePromotedType value;
+    if (!absl::SimpleAtoi(str, &value)) return std::nullopt;
+    if constexpr (kBits <= 16) {
+      if (value > GetMax()) {
+        LOG(ERROR) << "value(" << str
+                   << ") is greater than or equal to modulus";
+        return std::nullopt;
+      }
+    }
+    return PrimeField(value);
+  }
+  static std::optional<PrimeField> FromHexString(std::string_view str) {
+    using MaybePromotedType = std::conditional_t<(kBits <= 32), uint32_t, Type>;
+    MaybePromotedType value;
+    if (!absl::SimpleHexAtoi(str, &value)) return std::nullopt;
+    if constexpr (kBits <= 16) {
+      if (value > GetMax()) {
+        LOG(ERROR) << "value(" << str
+                   << ") is greater than or equal to modulus";
+        return std::nullopt;
+      }
+    }
+    return PrimeField(value);
+  }
+
+  constexpr static PrimeField FromBigInt(BigInt<N> big_int) {
+    return PrimeField(big_int);
+  }
+
+  static PrimeField FromMpzClass(const mpz_class& value) {
+    BigInt<N> big_int;
+    gmp::CopyLimbs(value, big_int.limbs);
+    return FromBigInt(big_int);
+  }
+
+  static void Init() { VLOG(1) << Config::kName << " initialized"; }
+
+  constexpr value_type value() const { return value_; }
+
+  constexpr bool IsZero() const { return value_ == 0; }
+
+  constexpr bool IsOne() const { return value_ == 1; }
+
+  std::string ToString() const { return base::NumberToString(value_); }
+
+  std::string ToHexString(bool pad_zero = false) const {
+    std::string str = base::HexToString(value_);
+    if (pad_zero) {
+      str = base::ToHexStringWithLeadingZero(str, 8);
+    }
+    return base::MaybePrepend0x(str);
+  }
+
+  mpz_class ToMpzClass() const {
+    mpz_class ret;
+    gmp::WriteLimbs(ToBigInt().limbs, N, &ret);
+    return ret;
+  }
+
+  constexpr BigInt<N> ToBigInt() const { return BigInt<N>(value_); }
+
+  constexpr operator Type() const { return value_; }
+
+  Type operator[](size_t i) const {
+    DCHECK_EQ(i, size_t{0});
+    return value_;
+  }
```

Comment:
Please remove this.

Reply:
Is there a reason why we should remove it?

Reply:
Because we already have value() method? I don't get where this operator is being used for.

Reply:
@chokobole Is this for casting as well??

Reply:
Oh I see. it's for the compatibility with other field implementations! But I know what you mean! Can we proceed this after internal discussion since PrimeFields for small modulus already have this.

Reply:
Okie!